contracts/modules/cover/Cover.sol

```diff
@@ -173,124 +191,111 @@ contract Cover is ICover, MasterAwareV2 {
+        if (coveredAmountInNXM < requestedCoverAmountInNXM && i + 1 < params.coverChunkRequests.length) {
+          params.coverChunkRequests[i + 1].coverAmountInAsset +=
+          (requestedCoverAmountInNXM - uint96(coveredAmountInNXM)) * tokenPrice / 1e18;
         }
+      }
 
-        amountLeftToCoverInNXM -= coveredAmountInNXM;
-        totalPremiumInNXM += premiumInNXM;
-
-        {
-
-          for (uint j = 0; j < originalCoverChunks.length; j++) {
-            if (originalCoverChunks[j].poolAddress == coverChunkRequests[i].poolAddress) {
-              // if the pool already exists add the previously existing amounts
-              coveredAmountInNXM = coveredAmountInNXM + originalCoverChunks[j].coverAmountInNXM;
-              // set the premium as the premium remainder for the rest of the period + the newly paid premium
-              premiumInNXM = premiumInNXM + originalCoverChunks[j].premiumInNXM * remainingPeriod / originalCover.period;
-              break;
-            }
+      totalPremiumInNXM += premiumInNXM;
+      totalCoverAmountInNXM += coveredAmountInNXM;
+      {
+
+        for (uint j = 0; j < originalCoverChunks.length; j++) {
+          if (originalCoverChunks[j].poolAddress == params.coverChunkRequests[i].poolAddress) {
+            // if the pool already exists add the previously existing amounts
+            coveredAmountInNXM = coveredAmountInNXM + originalCoverChunks[j].coverAmountInNXM;
+            // set the premium as the premium remainder for the rest of the period + the newly paid premium
+            premiumInNXM = premiumInNXM + originalCoverChunks[j].premiumInNXM * remainingPeriod / originalCover.period;
+            break;
           }
-          coverChunksForCover[newCoverId].push(
-            CoverChunk(
-              coverChunkRequests[i].poolAddress,
-              uint96(coveredAmountInNXM),
-              uint96(premiumInNXM)
-            ));
         }
+        coverChunksForCover[newCoverId].push(
+          CoverChunk(
+            params.coverChunkRequests[i].poolAddress,
+            uint96(coveredAmountInNXM),
+            uint96(premiumInNXM)
+          ));
       }
-      require(amountLeftToCoverInNXM == 0, "Not enough available capacity");
     }
 
-    premiumInAsset = totalPremiumInNXM * tokenPrice / 1e18;
+    premiumInPaymentAsset = totalPremiumInNXM * pool().getTokenPrice(params.paymentAsset) / 1e18;
 
     // make the previous cover expire at current block
     uint32 elapsedPeriod = originalCover.period - remainingPeriod;
-    uint96 updatedOriginalPremium = originalCover.premium * elapsedPeriod / originalCover.period;
-    uint96 carriedPremium = originalCover.premium - updatedOriginalPremium;
+    uint96 updatedOriginalPremium =
+      uint96(originalCover.premium * originalCover.amount / BASIS_PRECISION * elapsedPeriod / originalCover.period);
+    uint96 carriedPremium = originalCover.premium * originalCover.amount - updatedOriginalPremium;
```

Comment:
s/carriedPremium/newCoverPremium

To be consistent with the variable naming above.